### PR TITLE
Implement moving blocks to previous/next page using context menu

### LINKED ISSUE #3330: Add option to move to previous/next page from context menu
## Feature description

As a user I want to be able to move elements to next or previous page via the context menu for said element from the right-click menu and from the existing context menu.

Related to #3311.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- This will move the element to the same relative position on the page, just on the desired page.
- This will be the only way to move Page Attachments to a different page, as they aren't draggable by themselves.
- Moving to previous page should only be possible for elements on page 2 and further along in the story (except for CTA elements - they can only be moved to previous if they're at least on page 3).
- Moving to next page should only be possible for elements on the penultimate page or earlier in the story.
- Some elements can only exist at most once on a page (Page Attachments and CTA's) - they shouldn't be allowed to be moved to pages that already have a similar item.
- Unavailable options (as per the above restrictions) should be hidden from the menu (or dimmed? but normally we hide in the UI as per current UX).

## Implementation brief

* 

### Test 1
1. Create 2 pages (They should be 3 pages in total)
1. Navigate to Page 1
1. Create text element (or reuse existing one)
1. Select text element
1. Right click on text element. 
1. Select 'Send to next page' (note the send to previous page should not be present). 
1. After element has moved, if not selected, select text element. 
1. Right click on text element. 
1. Select 'Send to previous page' (note the send to previous page and next page should be present). 

### Test 2
1. Create 3 pages (They should be 4 pages in total)
1. Navigate to Page 2
1. Create CTA element
1. Select CTA element
1. Right click on text element. 
1. Select 'Send to next page' (note the send to previous page should not be present). 
1. After element has moved, if not selected, select CTA element. 
1. Right click on CTA element. 
1. Select 'Send to previous page' (note the send to previous page and next page should be present).

## Demo

* https://youtu.be/JTEvudZHBsM

## Changelog entry

* Added support for moving elements across pages via keyboard

### PR DESCRIPTION
**Pull request by spacedmonkey**
Add option to move block back and forward depending using context menu.

Fixes #3330
Blocked by #3311 

This PR delays on several pieces of code found in #3311. Why testing, I copied some of this code into the PR, but once that PR is merged, this PR should be updated.

### CODE REVIEW

**Comment by spacedmonkey** [comment]
This PR may seem more complex than it is. But most of it just a refactor to use react hooks. 

This PR is heavily related to #3407 and we will need to decided which PR is should be merged first.

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 1]
```diff
@@ -100,16 +208,53 @@ const RightClickMenu = ( props ) => {
...
+		const pageNumber = pageList.length;
+		if ( block && pageNumber > 1 ) {
+			const currentPage = getCurrentPage();
+			const currentPagePosition = pageList.indexOf( currentPage );
+			if ( currentPagePosition > 0 ) {
+				const prevPage = getPageByOffset( -1 );
+				if ( isBlockAllowedOnPage( block.name, prevPage ) ) {
+					blockActions.push(
+						{
+							name: __( 'Send block to previous page', 'amp' ),
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]: While we're at it, I'd like to replace usage of the word "block" with "element" in this component. We should be more strict there.

```suggestion
							name: __( 'Move element to previous page', 'amp' ),
```

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 2]
```diff
@@ -100,16 +208,53 @@ const RightClickMenu = ( props ) => {
...
+						},
+					);
+				}
+			}
+			if ( currentPagePosition < ( pageNumber - 1 ) ) {
+				const nextPage = getPageByOffset( 1 );
+				if ( isBlockAllowedOnPage( block.name, nextPage ) ) {
+					blockActions.push(
+						{
+							name: __( 'Send block to next page', 'amp' ),
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]:
```suggestion
							name: __( 'Move element to next page', 'amp' ),
```

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 3]
```diff
@@ -100,16 +208,53 @@ const RightClickMenu = ( props ) => {
...
 					blockAction: duplicateBlock,
+					params: [ firstBlockClientId ],
 					icon: 'admin-page',
 					className: 'right-click-duplicate',
 				},
 			);
 		}
 
+		const pageList = getBlockOrder();
+		const pageNumber = pageList.length;
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]: What about `numPages` or `totalPages` or something? Was a bit confusing at first.

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 4]
```diff
@@ -100,16 +208,53 @@ const RightClickMenu = ( props ) => {
 				{
 					name: __( 'Duplicate Block', 'amp' ),
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]:
```suggestion
					name: __( 'Duplicate Element', 'amp' ),
```

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 5]
```diff
@@ -82,12 +188,14 @@ const RightClickMenu = ( props ) => {
 			{
 				name: __( 'Copy Block', 'amp' ),
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]:
```suggestion
				name: __( 'Copy Element', 'amp' ),
```

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 6]
```diff
@@ -82,12 +188,14 @@ const RightClickMenu = ( props ) => {
...
 				blockAction: copyBlock,
+				params: [ firstBlockClientId ],
 				icon: 'admin-page',
 				className: 'right-click-copy',
 			},
 			{
 				name: __( 'Cut Block', 'amp' ),
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]:
```suggestion
				name: __( 'Cut Element', 'amp' ),
```

File: assets/src/stories-editor/components/right-click-menu/index.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/stories-editor/components/right-click-menu/index.js]: As other items in the list already use block and that is like in 1.3, that is out of scope of this ticket. Please following up with another issue.

File: assets/src/stories-editor/components/right-click-menu/index.js (hunk 3 above)
Comment (spacedmonkey) [reply assets/src/stories-editor/components/right-click-menu/index.js]: Fixed in https://github.com/ampproject/amp-wp/pull/3417/commits/558739e2b25c4870320736e30689e719345b64f5

File: assets/src/stories-editor/components/right-click-menu/index.js [hunk 7]
```diff
@@ -100,16 +208,53 @@ const RightClickMenu = ( props ) => {
...
+			const currentPagePosition = pageList.indexOf( currentPage );
+			if ( currentPagePosition > 0 ) {
+				const prevPage = getPageByOffset( -1 );
+				if ( isBlockAllowedOnPage( block.name, prevPage ) ) {
+					blockActions.push(
+						{
+							name: __( 'Send block to previous page', 'amp' ),
+							blockAction: moveBlockToPage,
+							params: [ prevPage ],
+							icon: 'arrow-left-alt',
```
Comment (swissspidy) [inline assets/src/stories-editor/components/right-click-menu/index.js]: I'd use `select( 'core/block-editor' ).getSettings().isRTL` and switch the arrow icons when a RTL language is being used

Comment (spacedmonkey) [reply assets/src/stories-editor/components/right-click-menu/index.js]: So we support RTL ATM?

Comment (swissspidy) [reply assets/src/stories-editor/components/right-click-menu/index.js]: For the most part. RTL stylesheets are created automatically, which helps with the majority of things. There's currently an issue with the layout (see #3012), but that's no reason to neglect it.

Comment (spacedmonkey) [reply assets/src/stories-editor/components/right-click-menu/index.js]: RTL support - https://github.com/ampproject/amp-wp/pull/3417/commits/3527c8ad3c7bedc073ee1c37e43888704e294d46

**Comment by swissspidy** [comment]
@spacedmonkey Need to update the tests for `useMoveBlockToPage` now

**Pull request merged**